README.md
# Snowplow-Mini

[![Join the chat at https://gitter.im/snowplow/snowplow-mini](https://badges.gitter.im/snowplow/snowplow-mini.svg)](https://gitter.im/snowplow/snowplow-mini)
[![Build Status][build-image]][build-wf]
[![Release][release-image]][releases]
[![License][license-image]][license]

An easily-deployable, single instance version of Snowplow that serves three use cases:

1. Gives a Snowplow consumer (e.g. an analyst / data team / marketing team) a way to quickly understand what Snowplow "does" i.e. what you put in at one end and take out of the other
2. Gives developers new to Snowplow an easy way to start with Snowplow and understand how the different pieces fit together
3. Gives people running Snowplow a quick way to debug tracker updates (because they can)

## Features

* [x] Data is tracked and processed in real time
* [x] Added Iglu Server to allow for custom schemas to be uploaded
* [x] Data is validated during processing
  - This is done using both our standard Iglu schemas and any custom ones that you have loaded into the Iglu Server
* [x] Data is loaded into Elasticsearch
  - Can be queried directly or through a Kibana dashboard
  - Good and bad events are in distinct indexes
* [x] Create UI to indicate what is happening with each of the different subsystems (collector, enrich etc.), so as to provide developers a very indepth way of understanding how the different Snowplow subsystems work with one another

## Topology

Snowplow-Mini runs several distinct applications on the same box which are all linked by NSQ topics.  In a production deployment each instance could be an Autoscaling Group and each NSQ topic would be a distinct Kinesis Stream.

* Scala Stream Collector:
  - Starts server listening on `http://< sp mini public ip>/` which events can be sent to.
  - Sends "good" events to the `RawEvents` NSQ topic
  - Sends "bad" events to the `BadEvents` NSQ topic
* Stream Enrich:
  - Reads events in from the `RawEvents` NSQ topic
  - Sends events which passed the enrichment process to the `EnrichedEvents` NSQ topic
  - Sends events which failed the enrichment process to the `BadEvents` NSQ topic
* Elasticsearch Sink Good:
  - Reads events from the `EnrichedEvents` NSQ topic
  - Sends those events to the `good` Elasticsearch index
  - On failure to insert, writes errors to `BadElasticsearchEvents` NSQ topic
* Elasticsearch Sink Bad:
  - Reads events from the `BadEvents` NSQ topic
  - Sends those events to the `bad` Elasticsearch index
  - On failure to insert, writes errors to `BadElasticsearchEvents` NSQ topic

These events can then be viewed in Kibana at `http://< sp mini public ip>/kibana`.

![](https://raw.githubusercontent.com/snowplow/snowplow-mini/master/utils/topology/snowplow-mini-topology.jpg)

## Roadmap

* [ ] Support loading data into Redshift. To give analysts / data teams a good idea to understand what Snowplow "does".

## Documentation

Documentation is available at [our docs website][mini-docs].

## Vagrant

Some advice on how to handle certain errors if you're trying to build this locally with Vagrant.

### `The box 'ubuntu/xenial64' could not be found or could not be accessed in the remote catalog.`

Your Vagrant version is probably outdated. Use Vagrant 2.0.0+.

### `npm install` results in `enoent ENOENT: no such file or directory, open '/package.json'`

This is caused by trying to use NFS. Comment the relevant lines in `Vagrantfile`.

Most likely this will happen on `TASK [sp_mini_5_build_ui : Install npm packages based on package.json.]` but see also: https://discourse.snowplowanalytics.com/t/snowplow-mini-local-vagrant/2930.

## Copyright and license

Snowplow Mini is copyright 2016-2021 Snowplow Analytics Ltd.

Licensed under the **[Apache License, Version 2.0][license]** (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[mini-docs]: https://docs.snowplowanalytics.com/docs/open-source-components-and-applications/snowplow-mini/


[build-image]: https://github.com/snowplow/snowplow-mini/actions/workflows/publish.yml/badge.svg
[build-wf]: https://github.com/snowplow/snowplow-mini/actions/workflows/publish.yml

[release-image]: https://img.shields.io/github/v/release/snowplow/snowplow-mini?sort=semver&style=flat
[releases]: https://github.com/snowplow/snowplow-mini/releases

[license-image]: https://img.shields.io/badge/license-Apache--2-blue.svg?style=flat
[license]: https://www.apache.org/licenses/LICENSE-2.0


CHANGELOG
Version 0.13.0 (2021-12-03)
---------------------------
Upgrade axios to 0.21.2 (#313)
Upgrade ca-certificates upstream in the setup playbook (#312)
Make it possible to refresh Elasticsearch index mapping (#192)
Add dedicated endpoint to restart individual services (#310)
Bump Caddy to 2.4.6 (#303)
Bump cadvisor to 0.43.0 (#311)
Make Elasticsearch index creation fault-tolerant (#309)
Configure a base path for nsqadmin (close #217)
Bump NSQ to 1.2.1 (#300)
Add cronjob to crop nsq data (#297)
Add new endpoint to manage telemetry (#307)
Bump Stream Collector to 2.4.1 (#304)
Bump Stream Enrich to 2.0.3 (#305)
Bump Elastic stack to 6.8.20 (#306)
Fix invalid vagrant ip address (#302)

Version 0.12.6 (2021-05-11)
---------------------------
Fix typo in elasticsearch-loader-good image (#283)

Version 0.12.5 (2021-05-10)
---------------------------
Switch to google-github-actions/setup-gcloud (#272)
Bump Stream Enrich to 2.0.0 (#280)
Make vagrant setup noninteractive (#281)

Version 0.12.4 (2021-04-15)
---------------------------
Remove flaky integration test from CI (#279)
Update deprecated ansible features in ansible-go (#277)
Update UI to stop using Snowplow CDN (#249)
Update readme to provide step by step instructions on running locally (#273)
Migrate pulling of docker images to dockerhub (#276)
Switch paramiko installation to pip (#274)

Version 0.12.3 (2021-03-29)
---------------------------
Bump Stream Collector to 2.2.1 (#270)

Version 0.12.2 (2021-02-19)
---------------------------
Replace Gitter button with Discourse button (#261)
Bump Stream Collector to 2.1.2 (#263)

Version 0.12.1 (2021-02-04)
---------------------------
Extend copyright to 2021 (#264)
Bump base image to 18.04 (#265)

Version 0.12.0 (2020-12-15)
---------------------------
Fix cronjobs (#260)
Bump Stream Collector to 2.1.0 (#259)
Bump Elastic stack to 6.8.9 (#253)
Export logs to cloud logging services (#258)
Remove golang setup from CI/CD (#256)
Integrate cAdvisor (#254)

Version 0.11.0 (2020-11-04)
---------------------------
Migrate from Travis to GitHub Actions (#250)
Remove Authorization header (#245)
Update documentation on index page (#247)
Bump Enrich to 1.4.1 (#243)
Bump Stream Collector to 2.0.0 (#241)
Disable default redirect path in collector configuration (#239)
Add section about errors when building with Vagrant to README.md (#238)

Version 0.10.1 (2020-09-07)
---------------------------
Bump Enrich to 1.3.2 (#240)

Version 0.10.0 (2020-08-10)
---------------------------
Fix Quick Start Guide link in README (#237)
Bump Elasticsearch Loader to 1.0.0 (#234)
Bump Stream Enrich to 1.3.1 (#235)

Version 0.9.1 (2020-06-16)
--------------------------
Bump Stream Enrich to 1.1.3 (#232)

Version 0.9.0 (2020-05-25)
--------------------------
Bump Scala Stream Collector to 1.0.1 (#231)
Automate publicizing GCP deployment (#208)
Bump NSQ to 1.2.0 (#223)
Bump Stream Enrich to 1.1.0 (#229)

Version 0.8.1 (2020-04-16)
--------------------------
Make Iglu Server run in development mode (#227)

Version 0.8.0 (2020-03-23)
--------------------------
Bump Stream Enrich to 1.0.0 (#221)
Bump Elasticsearch Loader to 0.12.0 (#220)
Fix postgres not starting after image update (#219)
Configure collector to return SameSite=None; Secure cookie (#218)
Bump Scala Stream Collector to 1.0.0 (#206)
Bump Iglu Server to 0.6.1 (#205)

Version 0.7.0 (2020-01-20)
--------------------------
Disable IP lookup enrichment (#214)

Version 0.6.4 (2019-11-05)
--------------------------
Improve Kibana setup (#209)

Version 0.6.3 (2019-10-24)
--------------------------
Add basic auth to app endpoint (#203)

Version 0.6.2 (2019-07-12)
--------------------------
Fix LSB init script for caddy (#200)

Version 0.6.1 (2019-07-01)
--------------------------
Overhaul build process (#195)
Bump Stream Enrich to 0.21.0 (#198)
Remove the PII default enrichment (#193)
Add iglu central mirror to the resolver (#185)
Bump UI dependencies' versions (#196)
Fix enrichment upload mechanism (#194)
Fix initialization of dockerized services (#182)
Correct 0.6.0 release date in CHANGELOG (#179)
Bump base Ubuntu image to 16.04 (#190)

Version 0.6.0 (2018-08-07)
--------------------------
Bump Stream Enrich to 0.18.0 (#174)
Publish mini to bigger instance types (#173)
Add publication of image for Google Compute Engine (#36)
Add support for external Postgres for Iglu Server (#154)
Assemble the apps using Docker Compose (#23)
Extend copyright to 2018 (#164)
Add gitter badge (#157)
Install NTP (#167)
Add ami-factory credentials to .travis.yml (#143)
Fix typo in README (#160)
Expose nsqadmin UI (#144)
Update example user_data.sh (#148)
Ensure Stream Enrich resolver uses cacheTtl of one (#137)

Version 0.5.0 (2018-05-29)
--------------------------
Bump Packer to 1.2.3 (#163)
Bump Stream Enrich to 0.16.1 (#161)
Add log rotation for internal services to prevent out of disk issues (#151)
Update base Ubuntu AMI being used (#147)
Make build process of Caddy deterministic (#158)

Version 0.4.0 (2017-12-21)
--------------------------
Add Control Plane API call to write domain name to Caddyfile for automatic TLS (#118)
Add Control Plane API call to change username and password for basic HTTP authentication (#117)
Add Control Plane API call to add apikey for Iglu authentication (#116)
Add Control Plane API call to put your own custom enrichments in Snowplow Mini (#66)
Add Control Plane API call to add external Iglu schema registry to Snowplow Mini (#62)
Add Control Plane API call to return the Snowplow Mini version (#128)
Switch to using NSQ rather than named pipes (#24)
Bump JavaScript Tracker to 2.8.2 (#71)
Ensure UI links adhere to the current protocol being used (#127)
Show Control Plane notifications in-page using react-alert (#119)
Create index mappings on bootstrap (#104)
Use Caddy instead of Nginx for serving the Snowplow Mini dashboard (#130)
Add libffi-dev, libssl-dev, python-dev and markupsafe as dependencies (#133)
Update build process to build Caddy from source (#132)

.github/workflows/deploy.sh
#!/bin/bash
set -e

cd $GITHUB_WORKSPACE

###### input validation ######
if [ "$#" -ne 1 ]
then
	echo "Inappropriate number of arguments! Aborted."
	echo "Usage: deploy.sh tag"
	exit 1
else
	export SP_MINI_PLATFORM=`echo $1 | cut -d'/' -f2`
	export SP_MINI_SIZE=`echo $1 | cut -d'/' -f3`

	if [ "$SP_MINI_PLATFORM" == "aws" -o "$SP_MINI_PLATFORM" == "gcp" ]
	then
		echo "Platform recognized!"
	else
		echo "Unrecognized platform! Aborted."
		echo "Supported platforms: aws, gcp."
		exit 1
	fi

	if [ "$SP_MINI_SIZE" == "large" ]
	then
		# prepare env vars for docker-compose
		# to be used as -Xmx jvm option for Elasticsearch & Snowplow apps
		echo -n > provisioning/roles/docker/files/.env
		echo "ES_JVM_SIZE=4g" >> provisioning/roles/docker/files/.env
		echo "SP_JVM_SIZE=512m" >> provisioning/roles/docker/files/.env
		# prepare env var for packer
		# to be used to determine which instance type to use
		if [ "$SP_MINI_PLATFORM" == "aws" ]
		then
			export AWS_INSTANCE_TYPE=t2.large
		elif [ "$SP_MINI_PLATFORM" == "gcp" ]
		then
			export GCP_MACHINE_TYPE=n1-standard-2
		fi
	elif [ "$SP_MINI_SIZE" == "xlarge" ]
	then
		# prepare env vars for docker-compose
		# to be used as -Xmx jvm option for Elasticsearch & Snowplow apps
		echo -n > provisioning/roles/docker/files/.env
		echo "ES_JVM_SIZE=8g" >> provisioning/roles/docker/files/.env
		echo "SP_JVM_SIZE=1536m" >> provisioning/roles/docker/files/.env
		# prepare env var for packer
		# to be used to determine which instance type to use
		if [ "$SP_MINI_PLATFORM" == "aws" ]
		then
			export AWS_INSTANCE_TYPE=t2.xlarge
		elif [ "$SP_MINI_PLATFORM" == "gcp" ]
		then
			export GCP_MACHINE_TYPE=n1-standard-4
		fi
	elif [ "$SP_MINI_SIZE" == "xxlarge" ]
	then
		# prepare env vars for docker-compose
		# to be used as -Xmx jvm option for Elasticsearch & Snowplow apps
		echo -n > provisioning/roles/docker/files/.env
		echo "ES_JVM_SIZE=16g" >> provisioning/roles/docker/files/.env
		echo "SP_JVM_SIZE=3g" >> provisioning/roles/docker/files/.env
		# prepare env var for packer
		# to be used to determine which instance type to use
		if [ "$SP_MINI_PLATFORM" == "aws" ]
		then
			export AWS_INSTANCE_TYPE=t2.2xlarge
		elif [ "$SP_MINI_PLATFORM" == "gcp" ]
		then
			export GCP_MACHINE_TYPE=n1-standard-8
		fi
	else
		echo "Unrecognized size! Aborted."
		echo "Available sizes: large, xlarge and xxlarge."
		exit 1
	fi
fi

############## deployment ###################

if [ "$SP_MINI_PLATFORM" == "aws" ]
then
	echo -n > provisioning/resources/configs/compositions/.platform
	echo "aws" >> provisioning/resources/configs/compositions/.platform
	export AWS_ACCESS_KEY_ID=$AWS_DEPLOY_ACCESS_KEY
	export AWS_SECRET_ACCESS_KEY=$AWS_DEPLOY_SECRET_KEY
	packer build -only=amazon-ebs Packerfile.json
elif [ "$SP_MINI_PLATFORM" == "gcp" ]
then
	echo -n > provisioning/resources/configs/compositions/.platform
	echo "gcp" >> provisioning/resources/configs/compositions/.platform
	echo $GOOGLE_APPLICATION_CREDENTIALS_JSON_BASE64 | base64 --decode > $GITHUB_WORKSPACE/BIN
	export GOOGLE_APPLICATION_CREDENTIALS=$GITHUB_WORKSPACE/BIN
	packer build -only=googlecompute Packerfile.json
else
	echo "Unrecognized platform. Aborted."
	exit 1
fi


provisioning/resources/control-plane/change_credentials.go
/**
 * Copyright (c) 2016-2018 Snowplow Analytics Ltd.
 * All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache
 * License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.
 *
 * See the Apache License Version 2.0 for the specific language
 * governing permissions and limitations there under.
 */

package main

import (
	"encoding/base64"
	"errors"
	"io/ioutil"
	"strconv"
	"strings"

	"github.com/nbutton23/zxcvbn-go"
	"golang.org/x/crypto/bcrypt"
)

func changeCredentials(configPath string, username string, password string) (error, int) {

	hashedPassword, err := checkPassword(password)
	if err != nil {
		return err, 400
	}

	lines, err := fileToLines(configPath)
	if err != nil {
		return err, 500
	}
	fileContent := ""
	hit := false
	for _, line := range lines {
		if hit {
			line = "        " + username + " " + base64Encode(hashedPassword)
			hit = false
		}
		if strings.Contains(line, "basicauth") {
			hit = true
		}
		fileContent += line
		fileContent += "\n"
	}
	return ioutil.WriteFile(configPath, []byte(fileContent), 0644), 500
}

func checkPassword(password string) (string, error) {

	if len(password) < 8 {
		return "", errors.New("weak password: length can not be shorter than 8 characters")
	}

	minEntropyMatch := zxcvbn.PasswordStrength(password, nil)

	// see https://github.com/nbutton23/zxcvbn-go#use
	if minEntropyMatch.Score < 4 {
		return "", errors.New("weak password: strength score is " + strconv.Itoa(minEntropyMatch.Score) + " but must be 4 at least")
	}

	hashedPassword, err := bcryptHash(password)
	if err != nil {
		return "", err
	}

	return hashedPassword, nil
}

func bcryptHash(password string) (string, error) {
	hash, err := bcrypt.GenerateFromPassword([]byte(password), 31)
	if err != nil {
		return "", err
	}
	return string(hash), nil
}

func base64Encode(in string) string {
	return base64.StdEncoding.EncodeToString([]byte(in))
}


provisioning/resources/control-plane/main.go
/**
 * Copyright (c) 2016-2018 Snowplow Analytics Ltd.
 * All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache
 * License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.
 *
 * See the Apache License Version 2.0 for the specific language
 * governing permissions and limitations there under.
 */

package main

import (
	"errors"
	"flag"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"

	"github.com/BurntSushi/toml"
)

var configPath string
var config ControlPlaneConfig

func main() {
	configFlag := flag.String("config", "/home/ubuntu/snowplow/configs/control-plane-api.toml",
		"Control Plane API config file")
	flag.Parse()
	configPath = *configFlag

	if _, err := toml.DecodeFile(configPath, &config); err != nil {
		panic(err)
	}

	http.HandleFunc("/restart-services", restartServices)
	http.HandleFunc("/enrichments", uploadEnrichments)
	http.HandleFunc("/iglu-config", uploadIgluConfig)
	http.HandleFunc("/external-iglu", addExternalIgluServer)
	http.HandleFunc("/local-iglu-apikey", addLocalIgluApikey)
	http.HandleFunc("/credentials", changeUsernameAndPassword)
	http.HandleFunc("/domain-name", addDomainName)
	http.HandleFunc("/version", getSpminiVersion)
	http.HandleFunc("/telemetry", manageTelemetry)
	log.Fatal(http.ListenAndServe(":10000", nil))
}

func restartServices(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "PUT" {
		err := restartSPServices()
		if err != nil {
			http.Error(resp, err.Error(), 500)
		} else {
			resp.WriteHeader(http.StatusOK)
			io.WriteString(resp, "OK")
		}
	} else {
		// Return 404 for other methods
		http.Error(resp, "", 404)
	}
}

func uploadEnrichments(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		// maxMemory bytes of body's file parts are stored in memory,
		// with the remainder stored on disk in temporary files
		var maxMemory int64 = 32 << 20
		err := req.ParseMultipartForm(maxMemory)
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		file, handler, err := req.FormFile("enrichmentjson")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		defer file.Close()

		fileContentBytes, err := ioutil.ReadAll(file)
		fileContent := string(fileContentBytes)

		if !isJSON(fileContent) {
			http.Error(resp, "JSON is not valid", 400)
			return
		}

		f, err := os.OpenFile(config.Dirs.Enrichments+"/"+handler.Filename, os.O_WRONLY|os.O_CREATE, 0666)
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		defer f.Close()

		// Uploaded enrichment can be shorter than the existing one
		// Truncating to 0 bytes and seeking I/O offset to the beginning
		// Prevents the possibility of corrupted json
		f.Truncate(0)
		f.Seek(0, 0)
		// Now we can write to file in peace
		io.WriteString(f, fileContent)

		err = restartService("streamEnrich")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "uploaded successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func uploadIgluConfig(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		// maxMemory bytes of body's file parts are stored in memory,
		// with the remainder stored on disk in temporary files
		var maxMemory int64 = 32 << 20
		err := req.ParseMultipartForm(maxMemory)

		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		file, _, err := req.FormFile("igluserverhocon")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		defer file.Close()

		fileContentBytes, err := ioutil.ReadAll(file)
		fileContent := string(fileContentBytes)
		f, err := os.OpenFile(config.Dirs.Config+"/"+config.ConfigNames.IgluServer, os.O_WRONLY|os.O_CREATE, 0666)
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		defer f.Close()

		// Uploaded Iglu Server configuration can be shorter than existing one
		// Which would make iglu server configuration invalid
		// Truncating to 0 bytes and seeking I/O offset to the beginning
		// Prevents that possibility
		f.Truncate(0)
		f.Seek(0, 0)
		// Now we can write to config file in peace
		io.WriteString(f, fileContent)

		err = restartService("iglu")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "uploaded successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func addExternalIgluServer(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		req.ParseForm()

		vendorPrefixArr, checkVendor := req.Form["vendor_prefix"]
		uriArr, checkUri := req.Form["uri"]
		apikeyArr, checkApikey := req.Form["apikey"]
		nameArr, checkName := req.Form["name"]
		priorityArr, checkPriority := req.Form["priority"]
		if !(checkVendor && checkUri && checkName && checkPriority) {
			http.Error(resp, "missing parameter", 400)
			return
		}
		uri := uriArr[0]
		apikey := ""
		if checkApikey {
			apikey = apikeyArr[0]
		}
		vendorPrefix := vendorPrefixArr[0]
		name := nameArr[0]
		priority, err := strconv.Atoi(priorityArr[0])
		if err != nil {
			http.Error(resp, "Priority must be an integer", 400)
			return
		}

		if !isURLReachable(uri) {
			http.Error(resp, "Given URL is not reachable", 400)
			return
		}
		if apikey != "" && !isValidUUID(apikey) {
			http.Error(resp, "Given apikey is not a valid UUID.", 400)
			return
		}

		externalIgluServer := ExternalIgluServer{
			ConfigPath: config.Dirs.Config + "/" +
				config.ConfigNames.IgluResolver,
			IgluInfo: IgluInfo{
				VendorPrefix: vendorPrefix,
				Uri:          uri,
				Apikey:       apikey,
				Name:         name,
				Priority:     priority,
			},
		}

		err = externalIgluServer.addExternalIgluServer()
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		err = restartService("streamEnrich")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "added successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func addLocalIgluApikey(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		req.ParseForm()

		igluApikeyArr, checkApikey := req.Form["local_iglu_apikey"]
		if !checkApikey {
			http.Error(resp, "missing parameter", 400)
			return
		}
		igluApikey := igluApikeyArr[0]

		if !isValidUUID(igluApikey) {
			http.Error(resp, "Given apikey is not valid UUID", 400)
			return
		}

		psqlInfos := PsqlInfos{
			User:     config.Psql.User,
			Password: config.Psql.Password,
			Database: config.Psql.Database,
			Addr:     config.Psql.Addr,
		}

		localIglu := LocalIglu{
			ConfigPath: config.Dirs.Config + "/" +
				config.ConfigNames.IgluResolver,
			IgluApikey: igluApikey,
			Psql:       psqlInfos,
		}

		err := localIglu.addApiKey()
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		err = restartService("streamEnrich")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "added successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func changeUsernameAndPassword(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		req.ParseForm()

		newUsernameArr, checkUsername := req.Form["new_username"]
		newPasswordArr, checkPassword := req.Form["new_password"]
		if !(checkUsername && checkPassword) {
			http.Error(resp, "missing parameter", 400)
			return
		}
		newUsername := newUsernameArr[0]
		newPassword := newPasswordArr[0]

		err := changeCredentials(
			config.Dirs.Config+"/"+config.ConfigNames.Caddy,
			newUsername,
			newPassword,
		)
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		err = restartService("caddy")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "changed successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func addDomainName(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "POST" {
		req.ParseForm()

		domainNameArr, checkDomainName := req.Form["domain_name"]
		if !checkDomainName {
			http.Error(resp, "missing parameter", 400)
			return
		}
		domainName := domainNameArr[0]

		err := checkHostDomainName(domainName)
		if err != nil {
			http.Error(resp, err.Error(), 405)
			return
		}

		err = changeDomainName(
			config.Dirs.Config+"/"+config.ConfigNames.Caddy,
			domainName,
		)
		if err != nil {
			http.Error(resp, err.Error(), 405)
			return
		}

		err = restartService("caddy")
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}

		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, "added successfully")
	} else {
		http.Error(resp, "", 404)
	}
}

func getSpminiVersion(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "GET" {
		versionBytes, err := ioutil.ReadFile(config.VersionFilePath)
		if err != nil {
			http.Error(resp, err.Error(), 500)
			return
		}
		versionStr := string(versionBytes)
		resp.WriteHeader(http.StatusOK)
		io.WriteString(resp, versionStr)
	} else {
		http.Error(resp, "", 404)
	}
}

func manageTelemetry(resp http.ResponseWriter, req *http.Request) {
	if req.Method == "GET" {
		telemetryDisable, err := getTelemetryDisable(config.Dirs.Config + "/" + config.ConfigNames.Collector)
		handleError(resp, err, http.StatusInternalServerError)

		resp.WriteHeader(http.StatusOK)
		_, err = io.WriteString(resp, telemetryDisable)
		handleError(resp, err, http.StatusInternalServerError)
	} else if req.Method == "PUT" {
		err := req.ParseForm()
		handleError(resp, err, http.StatusInternalServerError)

		disable := req.PostFormValue("disable")
		if disable == "" {
			handleError(resp, errors.New("no key named disable"), http.StatusBadRequest)
		} else if disable == "true" || disable == "false" {
			err = setTelemetryDisable(config.Dirs.Config+"/"+config.ConfigNames.Collector, disable)
			handleError(resp, err, http.StatusInternalServerError)
			err = restartService("streamCollector")
			handleError(resp, err, http.StatusInternalServerError)
		} else {
			handleError(resp, errors.New("set disable key to either true or false"), http.StatusBadRequest)
		}
	} else {
		handleError(resp, errors.New("http method not supported"), http.StatusMethodNotAllowed)
	}
}

func handleError(w http.ResponseWriter, err error, status int) {
	if err != nil {
		http.Error(w, err.Error(), status)
		return
	}
}
